### PR TITLE
feat: List 2.0

### PR DESCRIPTION
**Pull request by e11sy**
This PR is related to the #75 issue
This PR is a part of global work on List 2.0

## Problems
- Can not convert from list to check list
- Data model has no place for meta (e.g. checked field for checklist)
- Nesting is not customizable 
- Can not increase nesting more than 1 time
```
1
    1.1.1
```
is currently unavailable
- Can't customize ordered list (custom numeration or starting with custom number)
## Solutions
- Separated rendering and tabulation for greater compliance with the SPR model
- Now ones list style would be changed, list would be rerendered
- Now data model will contain meta field
```
data : {
  items : [
    {
      "content": "Canon",
      "meta" : {
        checked: false
       },
      "items": [
        {
          "content": "Fisheye",
          "meta" : {
            checked: true
          },
          "items": [
            {
              "content": "Canon 15mm f/2.8",
              "meta" : {
                checked: true
              },
              "items": []
            },
          ]
        }
      ]
    } 
  ]
}   
```
- Nesting customization by `maxLevel` config option
- Add support for `<ol>` attributes in config (such as `type`, `start`, `reversed`)
## Status
- [x] Separate rendering and tabulation
- [x] Support checklist 
- [x] List rerendering on style change
- [x] Support new data model
- [x] Fix list bullets alignment
- [ ] Add nesting customization
- [ ] Add ordered list customization for #66
- [ ] Refactor paste config and nesting behaviour
- [ ] Export all available lists into toolbox for #70
## Additional
- Added `index.html` file and updated `vite.config.js`, so now dev mode is available
- Updated `@codexteam/icons` package to latest for checklist icon

### I will make ordered list and nesting customizations, refactoring of the paste config and nesting behaviour in the following pr

### CODE REVIEW

File: src/ListRenderer/checklistRenderer.ts
Comment (neSpecc) [inline src/ListRenderer/checklistRenderer.ts:1]: lets name files like classes from capital letter

File: src/ListRenderer/checklistRenderer.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+export class CheckListRenderer extends ListRenderer {
+  /**
+   * Tool's configuration
+   */
+  protected config?: NestedListConfig;
+
+  /**
+   * Is NestedList Tool read-only option
+   */
+  readOnly: boolean;
```
Comment (neSpecc) [inline src/ListRenderer/checklistRenderer.ts:19]: is it private?

File: src/ListRenderer/checklistRenderer.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+  /**
+   * Renders ol wrapper for list
+   * @returns - created html ol element
+   */
+  renderWrapper(): HTMLOListElement {
+    const listWrapper = Dom.make('ul', [ListRenderer.CSS.wrapper, ListRenderer.CSS.wrapperChecklist]) as HTMLOListElement;
+
+    listWrapper.addEventListener('click', (event) => {
+      this.toggleCheckbox(event);
```
Comment (neSpecc) [inline src/ListRenderer/checklistRenderer.ts:35]: move event delegation out for toggleCheckbox() to make it more straightforward

File: src/types/itemMeta.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Meta information of each list item
+ */
+export default interface ItemMeta {
```
Comment (neSpecc) [inline src/types/itemMeta.ts:4]: it should be CheckListItemMeta extending ItemMeta

File: src/ListRenderer/checklistRenderer.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+    return divElement;
+  }
+
+  /**
+   * Redners list item element
+   * @param content - content of the list item
+   * @returns - created html list item element
+   */
+  renderItem(content: string, meta: ItemMeta): HTMLLIElement {
```
Comment (neSpecc) [inline src/ListRenderer/checklistRenderer.ts:56]:
```suggestion
  renderItem(content: string, meta: CheckListItemMeta): HTMLLIElement {
```

File: src/ListRenderer/checklistRenderer.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+    return contentNode.innerHTML;
+  }
+
+  /**
+   * Return meta object of certain element
+   * @param {Element} item - item of the list to get meta from
+   * @returns {ItemMeta} Item meta object
+   */
+  getItemMeta(item: Element): ItemMeta {
```
Comment (neSpecc) [inline src/ListRenderer/checklistRenderer.ts:105]:
```suggestion
  getItemMeta(item: Element): CheckListItemMeta {
```

File: src/ListRenderer/listRenderer.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  wrapperChecklist: string;
+  item: string;
+  itemBody: string;
+  itemContent: string;
+  itemChildren: string;
+  settingsWrapper: string;
+  itemChecked: string;
+  noHover: string;
+  checkbox: string;
+  checkboxContainer: string;
```
Comment (neSpecc) [inline src/ListRenderer/listRenderer.ts:17]: list renderer should not know about checklist classes

File: src/ListRenderer/listRenderer.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
+/**
+ * CSS classes for the Nested List Tool
+ */
+interface NestedListCssClasses {
+  wrapper: string;
+  wrapperOrdered: string;
+  wrapperUnordered: string;
+  wrapperChecklist: string;
```
Comment (neSpecc) [inline src/ListRenderer/listRenderer.ts:8]: and here: list renderer should not know about children classes

File: src/ListRenderer/orderedListRenderer.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  /**
+   * Returns item meta, for ordered list checked will be always undefined
+   * @returns Item meta object
+   */
+  getItemMeta(): ItemMeta  {
+    return {
+      checked: undefined,
+    }
```
Comment (neSpecc) [inline src/ListRenderer/orderedListRenderer.ts:89]:
```suggestion
    return {
    }
```

File: src/ListRenderer/orderedListRenderer.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+    return contentNode.innerHTML;
+  }
+
+  /**
+   * Returns item meta, for ordered list checked will be always undefined
+   * @returns Item meta object
+   */
+  getItemMeta(): ItemMeta  {
```
Comment (neSpecc) [inline src/ListRenderer/orderedListRenderer.ts:86]:
```suggestion
  getItemMeta(): OrderedListItemMeta  {
```

File: src/ListRenderer/unorderedListRenderer.ts [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+  }
+
+
+  /**
+   * Returns item meta, for undered list checked will be always undefined
+   * @returns Item meta object
+   */
+  getItemMeta(): ItemMeta  {
+    return {
+      checked: undefined,
```
Comment (neSpecc) [inline src/ListRenderer/unorderedListRenderer.ts:92]: same

File: src/ListTabulator/index.ts [hunk 11]
```diff
@@ -0,0 +1,782 @@
...
+   */
+  appendItems(items: ListItem[], parentItem: Element): void {
+    if (this.list !== undefined) {
+      items.forEach((item) => {
+        const itemEl = this.list!.renderItem(item.content, item.meta);
+
+        parentItem.appendChild(itemEl);
+
+        if (item.items.length) {
+          const sublistWrapper = this.list?.renderSublistWrapper()
```
Comment (neSpecc) [inline src/ListTabulator/index.ts:184]: you can call `renderWrapper()` and a pass `level` in there

File: styles/index.pcss [hunk 12]
```diff
@@ -38,18 +45,112 @@
...
+    width: var(--width-checkbox);
+    height: var(--height-checkbox);
+    display: flex;
+    align-items: center;
+    cursor: pointer;
+
+    svg {
+      opacity: 0;
+      height: 20px;
+      width: 20px;
```
Comment (neSpecc) [inline styles/index.pcss:79]: why not `--width-checkbox`?

File: styles/index.pcss [hunk 13]
```diff
@@ -4,12 +4,25 @@
...
   list-style: none;
+  --radius-border: 5px;
+  --checkbox-background: #fff;
+  --color-border: #C9C9C9;
+  --color-bg-checked: #369FFF;
+  --line-height: 1.57em;
+  --color-bg-checked-hover: #0059AB;
+  --color-tick: #fff;
+  --width-checkbox: 20px;
+  --height-checkbox: 20px;
```
Comment (neSpecc) [inline styles/index.pcss:15]: maybe `--size-checkbox`?

File: styles/index.pcss [hunk 14]
```diff
@@ -38,18 +45,112 @@
...
+        }
+      }
+    }
+    &-check {
+      cursor: pointer;
+      display: inline-block;
+      position: relative;
+      margin: 0 auto;
+      width: 20px;
+      height: 20px;
```
Comment (neSpecc) [inline styles/index.pcss:134]: use variable

File: src/ListRenderer/checklistRenderer.ts [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+import * as Dom from '../utils/dom';
+import { ListRenderer } from './ListRenderer';
+
+interface ChecklistCssClasses {
+  wrapper: string;
+  wrapperChecklist: string;
+  item: string;
+  itemBody: string;
+  itemContent: string;
+  itemChildren: string;
```
Comment (neSpecc) [inline src/ListRenderer/checklistRenderer.ts]: they belongs to ListRenderer

File: src/ListRenderer/ChecklistRenderer.ts [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+import * as Dom from '../utils/dom';
+import { ListRenderer } from './ListRenderer';
+
+interface ChecklistCssClasses {
+  wrapper: string;
+  wrapperChecklist: string;
+  item: string;
+  itemBody: string;
+  itemContent: string;
+  itemChildren: string;
```
Comment (neSpecc) [inline src/ListRenderer/ChecklistRenderer.ts:13]: they belongs to ListRenderer

File: src/ListRenderer/OrderedListRenderer.ts [hunk 17]
```diff
@@ -0,0 +1,111 @@
+import type { OrderedListItemMeta } from "../types/itemMeta";
+import { NestedListConfig } from "../types/listParams";
+import * as Dom from '../utils/dom';
+import { ListRenderer } from './ListRenderer';
+
+/**
+ * CSS classes for the Ordered list
+ */
+interface OrderedListCssClasses {
+  wrapper: string;
```
Comment (neSpecc) [inline src/ListRenderer/OrderedListRenderer.ts:10]: same

File: src/ListRenderer/OrderedListRenderer.ts [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+  wrapperOrdered: string;
+  item: string;
+  itemBody: string;
+  itemContent: string;
+  itemChildren: string;
+}
```
Comment (neSpecc) [inline src/ListRenderer/OrderedListRenderer.ts:16]: same

File: src/ListRenderer/UnorderedListRenderer.ts [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+import * as Dom from '../utils/dom';
+import { ListRenderer } from './ListRenderer';
+
+interface UnoderedListCssClasses {
+  wrapper: string;
+  wrapperUnordered: string;
+  item: string;
+  itemBody: string;
+  itemContent: string;
+  itemChildren: string;
```
Comment (neSpecc) [inline src/ListRenderer/UnorderedListRenderer.ts:12]: and here

File: styles/index.pcss [hunk 20]
```diff
@@ -4,32 +4,40 @@
...
+  --checkbox-background: #fff;
+  --color-border: #C9C9C9;
+  --color-bg-checked: #369FFF;
+  --line-height: 1.57em;
+  --color-bg-checked-hover: #0059AB;
+  --color-tick: #fff;
+  --size-checkbox: 20px;
 
   &__item {
     line-height: 1.6em;
```
Comment (neSpecc) [inline styles/index.pcss:17]: why this override is needed?

File: styles/index.pcss [hunk 21]
```diff
@@ -38,18 +46,112 @@
...
+  }
 
+  &__settings {
     .cdx-settings-button {
       width: 50%;
     }
   }
+
+  &__checkbox {
+    padding-top: 0.1em;
```
[the remark this answers is not in the document]
Comment (e11sy) [reply styles/index.pcss:71]: removed since it is vertically centered with grid

**Review by neSpecc — APPROVED** [review]

**Pull request merged**